// @mastra/deployer/dist/chunk-3GJSIS6L.cjs
'use strict';

var path = require('path');
var localPkg = require('local-pkg');
var url = require('url');

// src/build/utils.ts
function isDependencyPartOfPackage(dep, packageName) {
  if (dep === packageName) {
    return true;
  }
  return dep.startsWith(`${packageName}/`);
}
function getPackageName(id) {
  const parts = id.split("/");
  if (id.startsWith("@")) {
    return parts.slice(0, 2).join("/");
  }
  return parts[0];
}
async function getPackageRootPath(packageName, parentPath) {
  let rootPath;
  try {
    let options = void 0;
    if (parentPath) {
      if (!parentPath.startsWith("file://")) {
        parentPath = url.pathToFileURL(parentPath).href;
      }
      options = {
        paths: [parentPath]
      };
    }
    const pkg = await localPkg.getPackageInfo(packageName, options);
    rootPath = pkg?.rootPath ?? null;
  } catch (e) {
    rootPath = null;
  }
  return rootPath;
}
function getCompiledDepCachePath(rootPath, packageName) {
  return slash(path.join(rootPath, "node_modules", ".cache", packageName));
}
function slash(path) {
  const isExtendedLengthPath = path.startsWith("\\\\?\\");
  if (isExtendedLengthPath) {
    return path;
  }
  return path.replaceAll("\\", "/");
}
function rollupSafeName(name, rootDir) {
  const rel = path.relative(rootDir, name);
  let entry = slash(rel);
  entry = entry.replace(/^(\.\.\/)+/, "");
  entry = entry.replace(/^\/+/, "");
  entry = entry.replace(/^[A-Za-z]:\//, "");
  if (!entry) {
    entry = slash(path.basename(name));
  }
  return entry;
}

exports.getCompiledDepCachePath = getCompiledDepCachePath;
exports.getPackageName = getPackageName;
exports.getPackageRootPath = getPackageRootPath;
exports.isDependencyPartOfPackage = isDependencyPartOfPackage;
exports.rollupSafeName = rollupSafeName;
exports.slash = slash;
